### ISSUE TITLE
secretGenerator does not update resourceNames in Role object

### ISSUE THREAD

**Issue by mgoodness** [kubernetes-sigs/kustomize#365]
kustomize v1.0.8

Given `role.yaml` (snip):
```
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: kubernetes-dashboard
rules:
- apiGroups:
  - ""
  resources:
  - secrets
  resourceNames:
  - kubernetes-dashboard
  verbs:
  - delete
  - get
  - update
```
and `kustomization.yaml` (snip):
```
commonLabels:
  k8s-app: kubernetes-dashboard

namespace: kube-system

secretGenerator:
- name: kubernetes-dashboard
  commands:
    ...
  type: Opaque
```
I get Secret:
```
apiVersion: v1
data:
  ...
kind: Secret
metadata:
  creationTimestamp: null
  labels:
    k8s-app: kubernetes-dashboard
  name: kubernetes-dashboard-62g8mtd5kg
  namespace: kube-system
type: Opaque
```
but Role:
```
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  labels:
    k8s-app: kubernetes-dashboard
  name: kubernetes-dashboard
  namespace: kube-system
rules:
- apiGroups:
  - ""
  resourceNames:
  - kubernetes-dashboard <- should be kubernetes-dashboard-62g8mtd5kg
  resources:
  - secrets
  verbs:
  - delete
  - get
  - update
```

**Comment by Liujingfang1** [comment]
```
  resources:
  - secrets
  resourceNames:
  - kubernetes-dashboard
```
It seems that this type of resource name reference depends on two fields: `resources` and `resourceNames`. `resources` tells it is a secret resource and `resourceName` tells the name of the secret. This is different from kustomize namereference, where only one field is involved.

**Comment by Liujingfang1** [comment]
A similar case to this is serviceaccount referenced in a ClusterRoleBinding
```
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: clusterrolebinding
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: clusterrole
subjects:
- kind: ServiceAccount
  name: serviceaccount
  namespace: ingress
```
For this case, Kustomize has the fieldspec `subjects/name` without checking `subjects/kind`.
So we can add `rules/resourceNames` to the secret's namereference.

**Comment by mgoodness** [comment]
Would it make sense to allow vars to be used in`rules/resourceNames` of Roles & ClusterRoles? I'm also running into this issue for resources created using `namePrefix`.

**Comment by Liujingfang1** [comment]
@mgoodness  I think this case should be solved by name reference rather than var reference. Currently the missing part in namereference is that it only works for string field, not work for array.

**Comment by Liujingfang1** [comment]
Should be fixed by #592
